pkg/cmd/pulumi/state.go

```diff
@@ -177,7 +176,10 @@ func totalStateEdit(ctx context.Context, s backend.Stack, showPrompt bool, opts
 
 	// If the stack is already broken, don't bother verifying the integrity here.
 	if !stackIsAlreadyHosed {
-		contract.AssertNoErrorf(snap.VerifyIntegrity(), "state edit produced an invalid snapshot")
+		err = snap.VerifyIntegrity()
```

Comment:
@pgavlin @Frassle do either of you see any harm in this? We return an error and should proceed to shut down gracefully from this point.

Reply:
I think that it would be fine to do this. I think the code needs a little tweak to move the `if err != nil` below into this `if` statement.

Reply:
It's not ideal. I think the assert was here because we shouldn't be causing bad snapshots in the first place. In the case the user hit in https://github.com/pulumi/pulumi/issues/11271 they deleted a resource that other resources still depended on. I think we should either not be allowing those deletes or (more likely) ensuring all the dependency chains are cut when we delete something.

Reply:
I agree with Fraser. I think the assert is intended to demonstrate the existence of a fatal bug elsewhere in the code. If we made the panic go away, but still cause the process to crash, the error message will look cleaner to the user but we lose the stack trace information.

I think it's better to identify the cause of the invalid snapshot and remediate the violated assertion.

Reply:
> I agree with Fraser. I think the assert is intended to demonstrate the existence of a fatal bug elsewhere in the code. If we made the panic go away, but still cause the process to crash, the error message will look cleaner to the user but we lose the stack trace information.

I really don't like the idea of having intentional panics, but I'll admit I may be in the minority.

My concern is more that we potentially make things worse by panicking and leaving anything else the engine is doing in an inconsistent state. If we're mid-write or mid-operation anywhere else, we potentially introduce _more_ customer issues.